Photo download application using OTA

mniestroj opened the pull request:
There are 3 artifacts (all of them are optional) as part of release that the firmware expects to receive:
- `greeting` - when received, saved on FS at `/storage/greeting`; this file is then displayed using `LOG_HEXDUMP_INF()` during boot
- `background` - when received, saved on FS at `/storage/background`; this file is then showed on the screen using LVGL
- `main` - when received, saved to second application slot, which is then swapped with first application slot during next boot (regular Zephyr firmware upgrade logic).

Image for `background` can be prepared using https://lvgl.io/tools/imageconverter online converter. An 128-pixel wide image was tested with this sample. Following settings were used:
- LVGLv8
- Color format: CF_TRUE_COLOR_ALPHA
- Output format: Binary RGB565 Swap

mniestroj committed:
bump SDK to pull newest Zephyr and NCS

Use new SDK with updated Zephyr to 3.7.0 and NCS to 2.7.0.

Provide implementation of mcuboot_assert_handler() as a workaround for the
case when chain-loading application from MCUboot bootloader is not
enabled (CONFIG_BOOTLOADER_MCUBOOT=n) and still there is use of bootloader
APIs. This is needed only for Espressif targets when NOT building with
'--sysbuild' (only useful during testing).

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
boards: native_sim: provide overlay configurations

This board will be used as a demo for photo OTA artifact download.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
handle OTA manifest updates directly by application

So far 'fw_update' module was responsible for handling OTA manifest updates
and then scheduling firmware upgrade. Since that application will utilize
multi component OTA, it makes sense to move logic about what (artifact) and
how till be updated at the application side.

mniestroj committed:
handle 'greeting' release/artifact

This artifacts is mainly for testing purposes, as it will be displayed on
console during boot and is lightweight (compared to firmware or any photo)
to test with.

mniestroj committed:
calculate sha256 of 'greeting' artifact

sha256 hash will be used in order to decide whether new (different)
artifact needs to be downloaded as part of OTA or not. Such approach does
not require to keep track of version of the artifact as separate metadata,
while making sure that both the artifact and version are in sync.

mniestroj committed:
integrate 'Zephyr + LVGL' hello world sample

This integrates FW upgrade and LVGL Hello World sample in single
application, using single event loop.

mniestroj committed:
boards: native_sim: extend storage_partiton 1MiB size

So far only 16KiB was available. Since (virtual) flash node is defined as
2MiB total, use second 1MiB for application storage.

mniestroj committed:
handle 'background' OTA package and file

Assume it is LVGL compatible image and show it during boot.

Such image can be prepared using https://lvgl.io/tools/imageconverter LVGL
online converter using following parameters:
 * LVGL v8
 * Color format: CF_TRUE_COLOR_ALPHA
 * Output format: Binary RGB565 Swap

mniestroj committed:
update only when hash (sha256) is different than the one used

mniestroj committed:
boards: xiao_esp32s3: add overlays

Reconfigure /storage partition to be 1MiB in size to easily fit any OTA
artifacts.

mniestroj committed:
boards: native_sim: use 16-bit swapped LVGL colors

This makes it the same as xiao_esp32s3 with 'seeed_xiao_round_display'
shield.

mniestroj committed:
handle firmware upgrade using OTA blockwise transfer API

Add new blockwise download handler for firmware upgrade, so firmware will
be written to second application slot instead of filesystem.

mniestroj committed:
move LVGL to Kconfig

This will allow to disable LVGL on boards that do not support it, while
still being able to test all the other features provided by this
application.

Also, enable LVGL only when "zephyr,display" chosen node is configured, so
that using 'xiao_esp32s3' board without 'seeed_xiao_round_display' shield
works as well.

mniestroj committed:
make display and LVGL optional

mniestroj committed:
allow maximum 3 OTA components

This allows to upgrade:
 * firmware
 * background (photo)
 * greeting (simple text)

in a single release.

mniestroj committed:
compare version and hash, depending on which one is set

mniestroj committed:
reboot after at least one successful artifact download

Reboot after at least one of the artifacts was successfully downloaded
using blockwise transfer. This will make sure that such artifact will be
used during next boot.

mniestroj committed:
README: add instructions for running this example

szczys commented on src/main.c:
```diff
@@ -118,11 +120,93 @@ static void on_ota_manifest(struct golioth_client *client,
+    fs_close(&fp);
+
+    return GOLIOTH_OK;
+}
```
When a seek or write error occurs, `GOLIOTH_OK` is still returned. Should an error be returned instead? Does the partially-written file also need to be cleaned up?

One suggestion is to set `int err = GOLIOTH_OK` at the top, then at the bottom `return err;`

mniestroj replied:
There is:
```
    if (err) {
        return GOLIOTH_ERR_FAIL;
    }
```
which will be hit now in case of error.

szczys commented on src/main.c:
```diff
@@ -229,6 +229,84 @@ static int greeting_show(void)
+    if (ret < 0) {
+        LOG_ERR("Failed to read: %zd", ret);
+        err = ret;
+        goto background_close;
```
Does the buffer need to be freed in this error case?

mniestroj replied:
Fixed by swapping `background_close` with `buffer_free` goto labels.

szczys commented on src/main.c:
```diff
@@ -229,6 +229,84 @@ static int greeting_show(void)
+    if (ret != dirent.size) {
+        LOG_ERR("ret (%d) != dirent.size (%d)", (int) ret, (int) dirent.size);
+        err = -EIO;
+        goto background_close;
```
Does the buffer need to be freed in this error case?

szczys commented on src/main.c:
```diff
@@ -15,7 +16,8 @@ LOG_MODULE_REGISTER(example_template, LOG_LEVEL_DBG);
 #include <zephyr/kernel.h>
 
 /* Current firmware version; update in prj.conf or via build argument */
```
```suggestion
/* Current firmware version; update in VERSION file */
```

mniestroj replied:
Related: https://github.com/golioth/example-template/pull/3

szczys commented on src/main.c:
```diff
@@ -128,32 +128,48 @@ static void on_ota_manifest(struct golioth_client *client,
 }
 
-static int component_name_hash_cmp(const char *name, uint8_t hash[32])
+static int component_hash_cmp(struct component_desc *desc, const uint8_t hash[32])
+{
+    return memcmp(desc->hash, hash, 32);
+}
+
+static int component_version_cmp(struct component_desc *desc, const char *version)
```
I have a bit of trouble understanding when each of these will be used. Perhaps an update to the commit message would clarify?

sam-golioth commented on src/flash.c:
```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2021 Golioth, Inc.
```
```suggestion
 * Copyright (c) 2024 Golioth, Inc.
```

mniestroj replied:
Updated to `2021-2024` due to almost all code being copy-pasted from https://github.com/golioth/golioth-zephyr-sdk/blob/main/samples/dfu/src/flash.c

sam-golioth commented on src/flash.c:
```diff
@@ -0,0 +1,151 @@
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(example_template);
```
We should change the module name

mniestroj replied:
Done

sam-golioth commented on src/flash.h:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2021 Golioth, Inc.
```
```suggestion
 * Copyright (c) 2024 Golioth, Inc.
```

sam-golioth commented on src/flash.h:
```diff
@@ -0,0 +1,61 @@
+	/* empty */
+};
+
+static inline int flash_img_prepare(struct flash_img_context *flash)
```
To match the function below (or can use `flash` below).
```suggestion
static inline int flash_img_prepare(struct flash_img_context *ctx)
```

sam-golioth replied:
Although `ctx` is probably better, because `flash` is close to `flush`

mniestroj replied:
Updated.

sam-golioth commented on src/flash.h:
```diff
@@ -0,0 +1,61 @@
+#include <stddef.h>
+#include <stdint.h>
+
+struct flash_img_context {
+	/* empty */
+};
```
```suggestion
struct flash_img_context;
```
Where is the concrete definition for this struct?

mniestroj replied:
This fragment is true only for `CONFIG_BOOTLOADER_MCUBOOT=n` case. MCUboot is disabled when **sysbuild** is not used. This is true mostly for testing and with `native_sim` platform, which does not have support for bootloader builds and chain-loading application image.

Note: this file is copy-pasted from https://github.com/golioth/golioth-zephyr-sdk/blob/main/samples/dfu/src/flash.h (which is BTW why 2021 is the year in the copyright statement).

sam-golioth commented on src/main.c:
```diff
@@ -7,15 +7,30 @@
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(example_template, LOG_LEVEL_DBG);
```
We should change this module name

mniestroj replied:
Done

sam-golioth commented on src/main.c:
```diff
@@ -72,10 +87,361 @@ static int app_settings_register(struct golioth_client *client)
+        return GOLIOTH_ERR_FAIL;
+    }
+
+    err = fs_seek(&fp, block_idx * 1024, FS_SEEK_SET);
```
We'll need a better way to handle this once we start negotiating the block size, but for now, we can at least use the kconfig value instead of magic number.
```suggestion
    err = fs_seek(&fp, block_idx * GOLIOTH_BLOCKWISE_DOWNLOAD_BUFFER_SIZE, FS_SEEK_SET);
```

sam-golioth commented on src/main.c:
```diff
@@ -72,10 +87,361 @@ static int app_settings_register(struct golioth_client *client)
+buffer_free:
+    free(buffer);
+
+background_close:
```
These `goto`s should be in the opposite order. If the call to `fs_open()` on line 361 fails, we'll attempt to close a file pointer that wasn't opened.

mniestroj replied:
Nice catch! Fixed.

sam-golioth commented on src/main.c:
```diff
@@ -90,24 +456,112 @@ int main(void)
+
+                LOG_INF("Updating %s package", component->package);
+
+                status = golioth_blockwise_get(client, ".u", &component->uri[4], GOLIOTH_CONTENT_TYPE_OCTET_STREAM, desc->write_block, component->package);
```
We should use `golioth_ota_download_component()` instead of calling `golioth_blockwise_get()` directly.

mniestroj replied:
Converted to use `golioth_ota_download_component()`. Thanks for suggestion!

sam-golioth commented on west-ncs.yml:
```diff
@@ -7,7 +7,7 @@ manifest:
     - name: golioth
       path: modules/lib/golioth-firmware-sdk
-      revision: v0.14.0
+      revision: example-download-photo
```
We should be able to push this back to `v0.14.0` now. The only change on this branch that this PR needs is to expose `golioth_coap_blockwise_get()` but we should be using `golioth_ota_download_component()` instead.

mniestroj committed:
west-zephyr: drop 'path-prefix' and 'name-allowlist'

It makes more sense to be consistent with Zephyr upstream and DO NOT use
any path-prefix. This also allows to share the same cloned repositories
between upstream Zephyr (when used as west manifest repo) and downstream
application.

Drop 'name-allowlist' to use all possible repositories from Zephyr.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
rename as example-download-photo

This is the first commit starting "download photo" example. Update
documentation and code that referenced "template" to reference
"download-photo" instead.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

sam-golioth commented on west-ncs.yml:
```diff
@@ -7,7 +7,7 @@ manifest:
     - name: golioth
       path: modules/lib/golioth-firmware-sdk
-      revision: v0.14.0
+      revision: bump-zephyr-ncs
```
Do you need the latest for this example? We're not likely to release v0.15 before the launch, given the current blockers, so we should move this back to v0.14 if we can.

mniestroj replied:
This sample assumes use of `seeed_xiao_round_display`, which is available upstream only AFTER 3.6.0. So we need newer version that the one supported by SDK 0.14.0. `native_sim` use (for development purposes with virtual display, or even for customers to try out) is another reason to use newest Zephyr.

mniestroj committed:
prj: drop Golioth logging backend

This will not be used in the example.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
background: add Echo-Pose-Stand.bin

Add generated background image that is compatible with LVGL format.
